input methods: use GTK3_IM_MODULE_FILE

ericsagnes opened the pull request:
~~**Work in progress, please do not merge yet.**~~

This is an implementation of #14019 fix.

@ttuegel Could you review this PR code?

A few personal concerns:
- ~~Using a systemd unit to generate the cache files doesn't feel right, is there a better way to do that?~~ solved in the second commit
- ~~Having an option (`i18n.inputMethod.package`) just to pass the input method package doesn't feel right.~~ solved by setting `internal = true`

Note: 
- Testing this PR will trigger a quite long rebuild of the gtk+3 package.
- I add commits to keep track of progress, all commits will be merged into one before merging.

abbradar commented on nixos/modules/i18n/inputMethod/default.nix:
```diff
@@ -24,6 +28,41 @@ with lib;
+      serviceConfig = {
+        Type = "oneshot";
+        Environment = ''
+          "GTK_PATH=${cfg.package}/lib/gtk-2.0/:${cfg.package}/lib/gtk-3.0/"
```
Use `Environment` as a set: `Environment.GTK_PATH = "...."`.

abbradar commented:
If `gtk-query-immodules-*` doesn't access anything but GTK_PATH then you can run it in a _package_. Then link results from the package into `/etc`. I have used this approach for example [here](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/hardware/udev.nix#L135).

ericsagnes commented:
@abbradar Thanks, I added a commit to remove the systemd unit and use `mkDerivation` instead. It looks cleaner now.

abbradar commented on nixos/modules/i18n/inputMethod/default.nix:
```diff
@@ -3,8 +3,20 @@
+    buildCommand = ''
+      GTK_PATH=${cfg.package}/lib/gtk-3.0/ gtk-query-immodules-3.0 > $out
+    '';
+  };
```
Add `preferLocalBuild = true; allowSubstitutes = false;` to those -- they look simple enough.

ericsagnes replied:
Thanks, I added a commit fixing this.

abbradar commented:
Nice! LGTM, I'll wait for someone else with more GTK expertise to review it but if I understand everything correctly, this should work. Also we may want to port this to stable although it's somewhat mass-rebuildy. @domenkozar, what do you think?

abbradar commented:
Oh, one more thing -- don't sweat over `package` option, I believe it's our standard way to do this -- just grep for `internal = true`.

abbradar commented on nixos/modules/i18n/inputMethod/default.nix:
```diff
@@ -24,6 +28,41 @@ with lib;
+
+      package = mkOption {
+        type    = types.path;
+        default = null;
```
Oh, right -- you should mark this as `internal = true`, I think.

ericsagnes replied:
Neat! Thanks, I added a commit.

ericsagnes commented:
Looking at previous discussions on the topic (#2450), I think that @lethalman might be interested by this PR.

ttuegel commented:
@ericsagnes Would you mind patching for the other environment variables GTK 3 reads, the same way you have done for `GTK_IM_MODULE_FILE`? It needs to be done eventually, and I would hate to incur multiple mass rebuilds.

ericsagnes commented:
@ttuegel No problems, looking at #5053 I only see `GTK_DATA_PREFIX`. Are there any other variables that should be patched?

ttuegel commented:
We really need to patch every call to `g_getenv("GTK_*")` because GTK makes no effort to determine if the listed paths are for GTK 2 or GTK 3.

ericsagnes commented:
@ttuegel Result of a quick `grep` in gtk3 source:

- `GTK_THEME`
- `GTK_DEBUG`
- `GTK_OVERLAY_SCROLLING`
- `GTK_CSS_DEBUG`
- `GTK_DATA_PREFIX`
- `GTK_EXE_PREFIX`
- `GTK_PATH`
- `GTK_STYLE_CONTEXT_WARNING`
- `GTK_IM_MODULE`
- `GTK_MODULES`
- `GTK_TEST_TOUCHSCREEN`
- `GTK_INSPECTOR_DISPLAY`
- `GTK_WIDGET_ASSERT_COMPONENTS`
- `GTK_CSD`

Is it fine to patch all these?

ttuegel commented:
Yes, that sounds good!

abbradar commented:
From cursory glance, some of these may be unnecessary (for example, `GTK_DEBUG` or `GTK_TEST_TOUCHSCREEN`). I'm not sure, however.

taku0 commented:
Tested using `nixos-rebuild build-vm` with a configuration.nix below:
https://gist.github.com/taku0/f988cd6c0f1151f8ee89bc7aa24bdcfb

I got a strange behavior: if `GTK_IM_MODULE_FILE` is defined and `~/.uim.d/customs/custom-*` does not exist, uim crashes with the following message:
```
Error: in scm_symbol_value: unbound variable: generic-commit-key?
```

The error does not occur with `master` branch (`GTK_IM_MODULE_FILE` is not defined) or if `~/.uim.d/customs/custom-*` are created with `uim-pref-gtk`. The error occurs again if `~/.uim.d/customs/custom-*` are removed. This could be a bug of uim or Anthy. I will investigate it later if possible.

ericsagnes commented:
@ttuegel I added a commit patching all the variables cited in the previous comment.
`gtk+3` is compiling fine, but I don't know how to check that all the new `GTK3*` variables are working correctly.

abbradar commented on pkgs/development/libraries/gtk+/env-var.patch:
```diff
@@ -0,0 +1,269 @@
+ 
+ #ifdef G_ENABLE_DEBUG
+-  env_string = g_getenv ("GTK_DEBUG");
++  env_string = (g_getenv ("GTK_DEBUG")) || (g_getenv ("GTK_DEBUG"));
```
Copy-paste mistake, maybe?

abbradar commented on pkgs/development/libraries/gtk+/env-var.patch:
```diff
@@ -0,0 +1,269 @@
+ 
+       name = g_getenv ("GTK_INSPECTOR_DISPLAY");
++      
++      const gchar *name3 = g_getenv ("GTK_INSPECTOR_DISPLAY");
```
Same, this should be "GTK3" I think.

ericsagnes replied:
oops, thanks for noticing :)

abbradar commented:
@ericbmerritt Very cool! Maybe we want to try upstreaming this (of course, out of scope of this PR)?

abbradar commented:
Ouch, sorry, auto-completion. I meant @ericsagnes

ttuegel commented:
@ericsagnes I agree now that the variables which are testing for `NULL`ity should not have `GTK3_` variants. It breaks the logic of these debug options, I think. Although, I'm not sure anything is harmed.

ericsagnes commented:
@abbradar Good suggestion, I filed a ticket [upstream](https://bugzilla.gnome.org/show_bug.cgi?id=764551).

> I agree now that the variables which are testing for NULLity should not have GTK3_ variants. It breaks the logic of these debug options, I think. Although, I'm not sure anything is harmed.

@ttuegel I can revert that if you think it is better, should I?

abbradar commented:
@ericbmerritt Thanks! Let's see what upstream thinks.

From another look (brief, though) those variables can be actually useful to have GTK3_ variants:

* GTK_IM_MODULE_FILE
* GTK_THEME
* GTK_DATA_PREFIX
* GTK_PATH
* GTK_IM_MODULE

Notice that I haven't included `GTK_EXE_PREFIX` because it's visible in the patch that GTK uses it with `/gtk-3.0` postfix, so it should be fine even when co-used with GTK2. Maybe I'm missing something, though.

Others are not really important and/or mostly for debugging, I think.

Notice that I don't have any knowledge on GTK topics and this is just me reading the patch and concluding what a variable does from its name!

ttuegel commented:
> From another look (brief, though) those variables can be actually useful to have GTK3_ variants:
> 
>     GTK_IM_MODULE_FILE
>     GTK_THEME
>     GTK_DATA_PREFIX
>     GTK_PATH
>     GTK_IM_MODULE
> 
> Notice that I haven't included GTK_EXE_PREFIX because it's visible in the patch that GTK uses it with /gtk-3.0 postfix, so it should be fine even when co-used with GTK2. Maybe I'm missing something, though.

I think the problem with GTK_EXE_PREFIX as things stand now is that it *must* be shared with GTK 2. I would prefer to have separate variables; in worst-case scenario, they always take the same value. Otherwise, I agree with you list of variables that need separate prefixes. Thanks for looking into this!

abbradar commented:
@ttuegel Hm, indeed, this would be more convenient (allows us to separate GTK2 and 3 configuration, enable their support separately etc). So I agree, let's patch `_EXE_PREFIX` too.

ericsagnes commented:
@abbradar I added a commit to remove the patches of the following variables:

- `GTK_DEBUG`
- `GTK_OVERLAY_SCROLLING`
- `GTK_CSS_DEBUG`
- `GTK_STYLE_CONTEXT_WARNING`
- `GTK_TEST_TOUCHSCREEN`
- `GTK_INSPECTOR_DISPLAY`
- `GTK_WIDGET_ASSERT_COMPONENTS`
- `GTK_CSD`

So now, the only patched variables are:

- `GTK_IM_MODULE_FILE`
- `GTK_THEME`
- `GTK_DATA_PREFIX`
- `GTK_PATH`
- `GTK_IM_MODULE`
- `GTK_EXE_PREFIX`

abbradar commented:
Thank you! This looks ready to me now. 
-- 
Nikolay.

ericsagnes committed:
gtk3: patch environment variables

This patch the following environment variables logic, enabling to use a
`GTK3_` version as an override:

- GTK_IM_MODULE_FILE
- GTK_THEME
- GTK_DATA_PREFIX
- GTK_PATH
- GTK_IM_MODULE
- GTK_EXE_PREFIX

Details at https://github.com/NixOS/nixpkgs/pull/14417

ericsagnes committed:
input methods: use GTK3_IM_MODULE_FILE

abbradar commented:
This is a mass rebuild I think, so I'll cherry-pick this to staging -- is this okay? I also think this should be ported to stable -- how do we usually proceed with mass-rebuildy patches to release? cc @domenkozar

ericsagnes commented:
I arranged the commits so the PR is ready to merge.

@taku0 Have you found the cause of the `uim` crash? I tested the PR with `uim` on a test machine and couldn't reproduce the issue.

domenkozar commented:
@abbradar we can eventually backport to 16.03, there's no staging for it. But for mass rebuilds, it should first be committed to master and tested for a few weeks. Then open a PR to backport to stable with arguments why it's feasible to do so.

abbradar commented:
@domenkozar Got it, thanks 

Let's wait for @taku0's response to ensure it isn't this patch's fault and merge.

taku0 commented:
It crashes when I try to input to an text field of `uim-pref-gtk` or `uim-pref-gtk3` but not crashes on `firefox-bin` or `emacs`. So it would be fine for daily use.

ericsagnes commented:
@taku0 I would like to check if the same happens on my test machine, can you tell which text field of `uim-gtk-pref` and the steps to reproduce the crash?

taku0 commented:
1. Download https://gist.github.com/taku0/f988cd6c0f1151f8ee89bc7aa24bdcfb to a temporary directory.
2. `git fetch --force upstream pull/14417/head:pull_14417` in nixpkgs directory.
3. `git checkout pull_14417 -b tmp`
4. `git rebase master` (optional)
5. `nixos-rebuild build-vm -I nixpkgs=~/src/nixpkgs -I nixos-config=./configuration.nix` where `~/src/nixpkgs` and `./configuration.nix` should be adjusted.
6. `./result/bin/run-nixos-vm`
7. Login as `alice` with password `alice`.
8. (Ignore KDE error message)
8. Open `konsole` from start menu.
9. `uim-pref-gtk`
10. Open “SKK 辞書” (SKK dictionaries) group.
11. Focus “個人辞書ファイル” (Personal dictionary file) text field.
12. Ensure Anthy is enabled (leftmost button on the uim toolbar)
12. Switch Anthy input mode (2nd button from the left on the uim toolbar) to “ひらがな” mode.
13. Type some words into the text field, e.g. `ka`.
14. Error messages is dumped to the konsole.

![screenshot](https://cloud.githubusercontent.com/assets/870673/14252494/93a57afc-fac2-11e5-84d3-3e8a49c2108a.png)

abbradar commented:
Seeing the upstream reaction on this patch, I wonder why won't we just patch our default paths into GTK instead? So it would expect IM cache in `/etc/gtk-*/immodules.cache`, have data prefix `/var/lib/current-system/sw` etc. by default.  What would the disadvantages be?

ttuegel commented:
> What would the disadvantages be?

GTK applications won't work outside NixOS. I don't have a problem with that; given upstream's obstructionist attitude, we should drop all GTK packages anyway. That's not going to happen, so I would say: go with the patch we already have.

abbradar commented:
> GTK applications won't work outside NixOS.

A-ha, indeed: I haven't considered that. Thanks for the clarification!

ericsagnes commented:
@taku0 Thanks for the detailed steps! I could reproduce the errors in the vm, but it is working fine on my physical test machine. 
It might be related to the vm, do you have a way to test on a physical machine to see if it is working any better there?

taku0 commented:
I will try to test on this weekend.

ericsagnes commented:
Update from the upstream bug report:

- `GTK_*` environment variables shouldn't be used and some of them are deprecated (`GTK_IM_MODULE_FILE` and `GTK_THEME`)
- Instead of patching environment variables, they prefer patches to fix default behavior of gtk; So it is very unlikely they will use this solution.

From the [bug report](https://bugzilla.gnome.org/show_bug.cgi?id=764551):

> I would be much more interested in patches that fix cache file locations like that or just make gtk work without the cache than I am in adding extra environment variables as a 'quick fix'.
> The long term vision for this is that we want to move away from loading modules for these things altogether. GTK+ should just use the preferred input method framework of the platform it is running on (for Linux, that would be ibus), not let the use switch between dozens of unmaintained and half-working alternatives. 

So shall we merge the PR an keep this patch until upstream fixes the default functionality or try a different approach?

abbradar commented:
I say let's merge -- while upstream fixes this, we want to have a working solution for our users. I don't think we as distribution maintainers need a better fix (of course, we as _users_ may want one, but that's another story). I also propose to move discussion about @taku0's problem to another issue. If noone objects, I'll hit the button in several days (feel free to beat me to it, of course!).

astsmtl commented:
Thank you, this PR fixed my UIM. :)

astsmtl commented:
...and broke pasystray. Backtrace shows that pasystray's  GTK+3 loads UIM plugin for _GTK+2_ and gets SIGABRT later. We need to test upgrade path for other GTK+3 apps before merging this.

ericsagnes commented:
@astsmtl I tried  `pasystray` with this PR and couldn't reproduce your issue.
I checked `pasystray` source to see if it was using any of the variables patched and couldn't find any reference.
Is there any special action that trigger the bug?

abbradar commented:
@astsmtl Wild guess -- maybe you use `pasystray` from `nix-env` without having it rebuilt? This patch would break all such software that uses old (unpatched) GTK+3.

taku0 commented:
> or try a different approach?

Another option is reverting a91161aa57f7cfe447a551bafc2d7c6c6bbbe44b partially for `gtkPlugins` [as ttuegel  suggests](https://github.com/NixOS/nixpkgs/issues/14019#issuecomment-201913378) (disclaimer: I'm the author of `gtkPlugins`: #3275), though I'm not against `GTK3_IM_MODULE_FILE`.

astsmtl commented:
Of course it was pasystray from my profile. But is this a good excuse for breaking? Maybe we can also patch GTK+2 to use GTK2_XXX variables?

ttuegel commented:
> Of course it was pasystray from my profile. But is this a good excuse for breaking?

This is a fair point.

One way we could fix this issue without breaking existing packages is to patch GTK 2 and 3 to search NIX_PROFILES for an immodules.cache file. I patched Qt 5 this way for its plugins. We could have GTK search each profile for a file `$PROFILE/etc/gtk-{2,3}.0/immodules.cache`. This simultaneously solves the backwards-compatibility problem, it allows users to override the system immodules.cache, and it works on non-NixOS systems.

taku0 commented:
Ideally, the following three GTK+ versions must match:

- A. the version used to build the application
- B. the version used to build the input method (or other plugins)
- C. the version used at runtime

The `GTK3_IM_MODULE_FILE` approach fixes B system-wide, so A-B and B-C may be different.
The `gtkPlugins` approach fixes B and C system-wide, so A-B and A-C may be different.
The wrapping-all-GTK+-applications approach suggested by @lethalman on #2450 fixes none of them, so all versions match, but requires to wrap all existing and future GTK+ applications. To make matters worse, if a non-privilege user installs an input method, it will effective only for applications installed by that user but not for system packages.

This is an unsolved problem of Nix.

taku0 commented:
@ericsagnes
It fails on my physical x86_64 machine.

`uim-pref-gtk`: fails
`uim-pref-gtk3`: fails
`uim-pref-qt4`: works
`firefox`: works
`emacs`: works
`inkscape`: works

ericsagnes commented:
> One way we could fix this issue without breaking existing packages is to patch GTK 2 and 3 to search NIX_PROFILES for an immodules.cache file. I patched Qt 5 this way for its plugins. We could have GTK search each profile for a file $PROFILE/etc/gtk-{2,3}.0/immodules.cache. This simultaneously solves the backwards-compatibility problem, it allows users to override the system immodules.cache, and it works on non-NixOS systems.

@ttuegel That sounds like a cleaner approach, I made an implementation [here](https://github.com/ericsagnes/nixpkgs/tree/fix/gtk-immodules-cache2). It is working fine with `gedit` (gtk3) and `geany` (gtk2). It trigger long mass rebuilds. 
This fixes #14019 but does nothing for #5053.
Should I make a new PR with the NIX_PROFILES version and close this one?

@taku0 thanks for testing, I have no idea about what can be the cause as other gtk apps are working. Does the NIX_PROFILES [version](https://github.com/ericsagnes/nixpkgs/tree/fix/gtk-immodules-cache2) work any better?

ttuegel commented:
> Should I make a new PR with the NIX_PROFILES version and close this one?

Yes, I think that would be preferable.

ericsagnes commented:
closing in favor of #14568.
